src/runtime/audience-action-local-flow.ts

```diff
@@ -161,11 +173,64 @@ export class AudienceActionLocalFlow implements AudienceActionFlow {
   private async initPrompt_() {
     if (this.params_.action === TYPE_REWARDED_AD) {
       await this.initRewardedAdWall_();
+    } else if (this.params_.action === TYPE_NEWSLETTER_SIGNUP) {
+      await this.initNewsletterSignup_();
+    } else {
+      this.renderErrorView_();
+    }
+  }
+
+  private async initNewsletterSignup_() {
+    this.eventManager_.logSwgEvent(
+      AnalyticsEvent.IMPRESSION_BYOP_NEWSLETTER_OPT_IN
+    );
+    const config = await this.getConfig_();
+    const codeSnippet = config?.optInParameters?.codeSnippet;
+
+    const validNewsletterSignupParams =
+      codeSnippet &&
+      config?.optInParameters?.promptPreference ===
+        PREFERENCE_PUBLISHER_PROVIDED_PROMPT;
+
+    if (validNewsletterSignupParams) {
+      this.renderOptinPrompt_(codeSnippet);
     } else {
+      this.eventManager_.logSwgEvent(
+        AnalyticsEvent.EVENT_BYOP_NEWSLETTER_OPT_IN_CONFIG_ERROR
+      );
       this.renderErrorView_();
     }
   }
 
+  private renderOptinPrompt_(codeSnippet: string) {
+    if (!codeSnippet.includes('form')) {
```

Comment:
Consider moving HTML validation to a separate method.

Reply:
Resolving this comment since this is deleted!